[quant] Make choose_qparams_optimized return Tensors to preserve dtype

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45530 [quant] Make choose_qparams_optimized return Tensors to preserve dtype**

Summary:
Returning double values requires special handling as a return type for aten functions.
Instead return tensors where the type is preserved in the tensor dtype

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_choose_qparams_optimized

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24001134](https://our.internmc.facebook.com/intern/diff/D24001134)

supriyar committed:
[quant] Make choose_qparams_optimized return Tensors to preserve dtype

Summary:
Returning double values requires special handling as a return type for aten functions.
Instead return tensors where the type is preserved in the tensor dtype

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_choose_qparams_optimized

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make choose_qparams_optimized return Tensors to preserve dtype"


Summary:
Returning double values requires special handling as a return type for aten functions.
Instead return tensors where the type is preserved in the tensor dtype

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_choose_qparams_optimized

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24001134](https://our.internmc.facebook.com/intern/diff/D24001134)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make choose_qparams_optimized return Tensors to preserve dtype"


Summary:
Returning double values requires special handling as a return type for aten functions.
Instead return tensors where the type is preserved in the tensor dtype

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_choose_qparams_optimized

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24001134](https://our.internmc.facebook.com/intern/diff/D24001134)

[ghstack-poisoned]

dskhudia commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4437,7 +4437,7 @@
   variants: function
 
-- func: choose_qparams_optimized(Tensor input, int numel, int n_bins, float ratio, int bit_width) -> (float, float)
+- func: choose_qparams_optimized(Tensor input, int numel, int n_bins, float ratio, int bit_width) -> (Tensor, Tensor)
```
I was wondering the reason for having numel as an argument? Isn't it the same as input tensor size?

supriyar replied:
Not necessarily, it could be per-channel in which case it will be numel per channel

supriyar committed:
Update on "[quant] Make choose_qparams_optimized return Tensors to preserve dtype"


Summary:
Returning double values requires special handling as a return type for aten functions.
Instead return tensors where the type is preserved in the tensor dtype

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_choose_qparams_optimized

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24001134](https://our.internmc.facebook.com/intern/diff/D24001134)

[ghstack-poisoned]
